File: custom_adapter/UnityLevelPlayAdapter/Objc/ISAlxCustomInterstitial.m
//
//  ISAlxCustomInterstitial.m
//  AlxAdsOCDemo
//
//  LevelPlay Interstitial 广告适配器
//

#import "ISAlxCustomInterstitial.h"
#import <AlxAds/AlxAds-Swift.h>

static NSString *const TAG = @"ISAlxCustomInterstitial";

@interface ISAlxCustomInterstitial () <AlxInterstitialAdDelegate>
@property (nonatomic, strong, nullable) AlxInterstitialAd *interstitialAd;
@property (nonatomic, weak, nullable) id<ISInterstitialAdDelegate> adDelegate;
@end

@implementation ISAlxCustomInterstitial

#pragma mark - ISBaseInterstitial

/// LevelPlay 请求加载插屏广告
- (void)loadAdWithAdData:(ISAdData *)adData delegate:(id<ISInterstitialAdDelegate>)delegate {
    NSLog(@"%@: loadAd", TAG);
    self.adDelegate = delegate;

    [ISAlxCustomAdapter initSdkWithAdData:adData];

    NSString *unitId = adData.configuration[@"unitid"];
    if (!unitId || unitId.length == 0) {
        NSString *msg = @"unitid is empty";
        NSLog(@"%@: error: %@", TAG, msg);
        [delegate adDidFailToLoadWithErrorType:ISAdapterErrorTypeInternal
                                     errorCode:ISAdapterErrorMissingParams
                                  errorMessage:msg];
        return;
    }

    NSLog(@"%@: loadAd unitid=%@", TAG, unitId);

    self.interstitialAd = [[AlxInterstitialAd alloc] init];
    self.interstitialAd.delegate = self;
    [self.interstitialAd loadAdWithAdUnitId:unitId];
}

/// LevelPlay 检查广告是否就绪
- (BOOL)isAdAvailableWithAdData:(ISAdData *)adData {
    BOOL ready = self.interstitialAd != nil && [self.interstitialAd isReady];
    NSLog(@"%@: isAdAvailable = %@", TAG, ready ? @"YES" : @"NO");
    return ready;
}

/// LevelPlay 展示插屏广告
- (void)showAdWithViewController:(UIViewController *)viewController
                          adData:(ISAdData *)adData
                        delegate:(id<ISInterstitialAdDelegate>)delegate {
    NSLog(@"%@: showAd", TAG);
    self.adDelegate = delegate;

    if (![self isAdAvailableWithAdData:adData]) {
        NSString *msg = @"ad is not ready";
        NSLog(@"%@: error: %@", TAG, msg);
        [delegate adDidFailToShowWithErrorCode:ISAdapterErrorInternal errorMessage:msg];
        return;
    }

    dispatch_async(dispatch_get_main_queue(), ^{
        [self.interstitialAd showAdWithPresent:viewController];
    });
}

#pragma mark - AlxInterstitialAdDelegate

- (void)interstitialAdLoad:(AlxInterstitialAd *)ad {
    NSLog(@"%@: interstitialAdLoad", TAG);
    [self.adDelegate adDidLoad];
}

- (void)interstitialAdFailToLoad:(AlxInterstitialAd *)ad didFailWithError:(NSError *)error {
    NSLog(@"%@: interstitialAdFailToLoad: %@", TAG, error.localizedDescription);
    [self.adDelegate adDidFailToLoadWithErrorType:ISAdapterErrorTypeInternal
                                        errorCode:error.code
                                     errorMessage:error.localizedDescription];
}

- (void)interstitialAdImpression:(AlxInterstitialAd *)ad {
    NSLog(@"%@: interstitialAdImpression", TAG);
    [self.adDelegate adDidOpen];
}

- (void)interstitialAdClick:(AlxInterstitialAd *)ad {
    NSLog(@"%@: interstitialAdClick", TAG);
    [self.adDelegate adDidClick];
}

- (void)interstitialAdClose:(AlxInterstitialAd *)ad {
    NSLog(@"%@: interstitialAdClose", TAG);
    [self.adDelegate adDidClose];
    self.interstitialAd = nil;
}

- (void)interstitialAdFailToShow:(AlxInterstitialAd *)ad didFailWithError:(NSError *)error {
    NSLog(@"%@: interstitialAdFailToShow: %@", TAG, error.localizedDescription);
    [self.adDelegate adDidFailToShowWithErrorCode:ISAdapterErrorInternal
                                     errorMessage:error.localizedDescription];
}

@end

File: custom_adapter/UnityLevelPlayAdapter/Swift/ISAlxCustomInterstitial.swift
//
//  ISAlxCustomInterstitial.swift
//  AlxAdsDemo
//
//  LevelPlay Interstitial 广告适配器
//  文档参考: https://docs.unity.com/zh-cn/grow/levelplay/sdk/ios/build-custom-adapter
//

import Foundation
import UIKit
import IronSource
import AlxAds

@objc(ISAlxCustomInterstitial)
public class ISAlxCustomInterstitial: ISBaseInterstitial {

    private static let TAG = "ISAlxCustomInterstitial"

    private var interstitialAd: AlxInterstitialAd?
    private weak var adDelegate: ISInterstitialAdDelegate?

    // MARK: - ISBaseInterstitial

    public override func loadAd(with adData: ISAdData, delegate: ISInterstitialAdDelegate) {
        NSLog("%@: loadAd", Self.TAG)
        self.adDelegate = delegate
        ISAlxCustomAdapter.initSdk(with: adData)

        guard let unitId = adData.configuration["unitid"] as? String, !unitId.isEmpty else {
            let msg = "unitid is empty"
            NSLog("%@: error: %@", Self.TAG, msg)
            (delegate as? ISAlxAdapterFailureReporter)?.adDidFailToLoad(
                withErrorType: 2,
                errorCode: ISAdapterErrors.missingParams.rawValue,
                errorMessage: msg)
            return
        }

        NSLog("%@: loadAd unitid=%@", Self.TAG, unitId)
        let ad = AlxInterstitialAd()
        ad.delegate = self
        self.interstitialAd = ad
        ad.loadAd(adUnitId: unitId)
    }

    public override func isAdAvailable(with adData: ISAdData) -> Bool {
        let ready = interstitialAd?.isReady() ?? false
        NSLog("%@: isAdAvailable = %@", Self.TAG, ready ? "YES" : "NO")
        return ready
    }

    public override func showAd(with viewController: UIViewController,
                                adData: ISAdData,
                                delegate: ISInterstitialAdDelegate) {
        NSLog("%@: showAd", Self.TAG)
        self.adDelegate = delegate

        guard isAdAvailable(with: adData) else {
            let msg = "ad is not ready"
            NSLog("%@: error: %@", Self.TAG, msg)
            (delegate as? ISAlxAdapterFailureReporter)?.adDidFailToShow(
                withErrorCode: 1000,
                errorMessage: msg)
            return
        }

        DispatchQueue.main.async { [weak self] in
            self?.interstitialAd?.showAd(present: viewController)
        }
    }
}

// MARK: - AlxInterstitialAdDelegate

extension ISAlxCustomInterstitial: AlxInterstitialAdDelegate {

    public func interstitialAdLoad(_ ad: AlxInterstitialAd) {
        NSLog("%@: interstitialAdLoad", Self.TAG)
        adDelegate?.adDidLoad()
    }

    public func interstitialAdFailToLoad(_ ad: AlxInterstitialAd, didFailWithError error: Error) {
        NSLog("%@: interstitialAdFailToLoad: %@", Self.TAG, error.localizedDescription)
        (adDelegate as? ISAlxAdapterFailureReporter)?.adDidFailToLoad(
            withErrorType: 2,
            errorCode: (error as NSError).code,
            errorMessage: error.localizedDescription)
    }

    public func interstitialAdImpression(_ ad: AlxInterstitialAd) {
        NSLog("%@: interstitialAdImpression", Self.TAG)
        adDelegate?.adDidOpen()
    }

    public func interstitialAdClick(_ ad: AlxInterstitialAd) {
        NSLog("%@: interstitialAdClick", Self.TAG)
        adDelegate?.adDidClick()
    }

    public func interstitialAdClose(_ ad: AlxInterstitialAd) {
        NSLog("%@: interstitialAdClose", Self.TAG)
        adDelegate?.adDidClose()
        interstitialAd = nil
    }

    public func interstitialAdFailToShow(_ ad: AlxInterstitialAd, didFailWithError error: Error) {
        NSLog("%@: interstitialAdFailToShow: %@", Self.TAG, error.localizedDescription)
        (adDelegate as? ISAlxAdapterFailureReporter)?.adDidFailToShow(
            withErrorCode: 1000,
            errorMessage: error.localizedDescription)
    }
}
